spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/pubsub/GcpPubSubEmulatorConfiguration.java

```diff
@@ -53,10 +53,4 @@ public TransportChannelProvider transportChannelProvider(GcpPubSubProperties gcp
 				.build();
 		return FixedTransportChannelProvider.create(GrpcTransportChannel.create(channel));
 	}
-
-	@Bean
-	@ConditionalOnMissingBean
-	public CredentialsProvider credentialsProvider() {
```

Comment:
M-m-m. That's was exactly the point of this `GcpPubSubEmulatorConfiguration`. 

I would say that it would be better for all other non-involved tests to *exclude* this `GcpPubSubEmulatorConfiguration`  from their auto-configuartions.

Reply:
I disagree. In my test app, I want to be able to use the emulator with other services. I might even want to test sending credentials to the emulator. In other words, emulator and credentials are unrelated.

Reply:
But we have here `@ConditionalOnMissingBean`! Why is that not enough if you create your own `CredentialsProvider` bean?

Reply:
The problem is the granularity of the autoconfig. Consider the `@AutoConfigureBefore(GcpContextAutoConfiguration.class)` in this class. If the emulator is used, the _whole_ context gets "polluted" with this bogus credentials provider, and then `GcpContextAutoConfiguration` can't create the good one anymore.